tools/clang/lib/SPIRV/AstTypeProbe.cpp

```diff
@@ -1215,5 +1215,89 @@ QualType getHLSLMatrixType(ASTContext &astContext, Sema &S,
       TemplateName(templateDecl), templateArgumentList, canonType);
 }
 
+bool isResourceOnlyStructure(QualType type) {
+  // Remove arrayness if needed.
+  while (type->isArrayType())
+    type = type->getAsArrayTypeUnsafe()->getElementType();
+
+  if (const auto *structType = type->getAs<RecordType>()) {
+    for (const auto *field : structType->getDecl()->fields()) {
+      // isResourceType does remove arrayness for the field if needed.
+      if (!isResourceType(field) &&
+          !isResourceOnlyStructure(field->getType())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return false;
+}
+
+bool isStructureContainingResources(QualType type) {
+  // Remove arrayness if needed.
+  while (type->isArrayType())
+    type = type->getAsArrayTypeUnsafe()->getElementType();
+
+  if (const auto *structType = type->getAs<RecordType>()) {
+    for (const auto *field : structType->getDecl()->fields()) {
+      auto fieldType = field->getType();
+      // Remove arrayness if needed.
+      while (fieldType->isArrayType())
+        fieldType = fieldType->getAsArrayTypeUnsafe()->getElementType();
```

Comment:
Can we remove this while statement based on the comment in the line 1225 `isResourceType does remove arrayness for the field if needed.`?